README.rst
=========
sainsmart
=========


.. image:: https://img.shields.io/pypi/v/sainsmart.svg
        :target: https://pypi.python.org/pypi/sainsmart
        :alt: PyPi

.. image:: https://img.shields.io/travis/vicyap/sainsmart.svg
        :target: https://travis-ci.org/vicyap/sainsmart
        :alt: Travis Build

.. image:: https://readthedocs.org/projects/sainsmart/badge/?version=latest
        :target: https://sainsmart.readthedocs.io/
        :alt: Documentation Status

.. image:: https://pyup.io/repos/github/vicyap/sainsmart/shield.svg
        :target: https://pyup.io/repos/github/vicyap/sainsmart/
        :alt: Updates

.. image:: https://coveralls.io/repos/github/vicyap/sainsmart/badge.svg?branch=master
        :target: https://coveralls.io/github/vicyap/sainsmart?branch=master
        :alt: Coverage

.. image:: https://ci.appveyor.com/api/projects/status/q1p33rqk0ewcwm29?svg=true
        :target: https://ci.appveyor.com/project/victor/sainsmart/branch/master
        :alt: Windows Build


sainsmart contains code for working with sainsmart products.

Install
-------

``pip install sainsmart``

Usage
-----
To use EthernetRelay in a project::

    from sainsmart import devices

    relay = devices.EthernetRelay()

    # Access the state of the relays
    # relay.state() is a list of bools
    relay_state = relay.state()

    # Turn on the relay at index 0.
    relay.turn_on(0)

    # Turn off the relay at index 0.
    relay.turn_off(0)

    # Toggle the state of relay 0.
    relay.toggle(0)

    # Turn on all relays.
    relay.all_on()

    # Turn off all relays.
    relay.all_off()

The SainSmart Ethernet Relay defaults ip address: 192.168.1.4/30000.
This can be changed through their web interface.

To use EthernetRelay with a different IP address::

    from sainsmart import devices

    relay = devices.EthernetRelay('http://192.168.44.100/30000')


Example
-------
Here is an example with a real device::

   >>> from sainsmart import devices
   >>> relay = devices.EthernetRelay()
   >>> relay.state()
   [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False]
   >>> relay.turn_on(0)
   >>> relay.state()
   [True, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False]
   >>> relay.toggle(0)
   >>> relay.state()
   [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False]
   >>> relay.all_on()
   >>> relay.state()
   [True, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
   >>> relay.turn_off(0)
   >>> relay.state()
   [False, True, True, True, True, True, True, True, True, True, True, True, True, True, True, True]
   >>> relay.all_off()
   >>> relay.state()
   [False, False, False, False, False, False, False, False, False, False, False, False, False, False, False, False] 


* Free software: MIT license
* Documentation: https://sainsmart.readthedocs.io.


Credits
---------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage



HISTORY.rst
=======
History
=======

0.3.2 (2018-05-05)
------------------
* Update the packages requests and typing with new versions

0.3.1 (2017-06-20)
------------------
* make the regex for getting the state of the relays more specific

0.3.0 (2017-06-14)
------------------
* add python 3.4 support

0.2.1 (2017-06-14)
------------------
* add install section to README
* fix coverage badge in README

0.2.0 (2017-06-14)
------------------
* python2.7 support
* renamed sainsmart.py to devices.py

0.1.1 (2017-06-12)
------------------
* Update usage and docs

0.1.0 (2017-06-12)
-------------------
* First release on PyPI.




tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for sainsmart."""


tests/test_devices.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `sainsmart` package."""


import unittest
import httmock
import re

from sainsmart import devices


_MOCK = None
_99_30000 = ' <small><a href="http://192.168.1.4/30000"></a></small><a href="http://192.168.1.4/30000/{}">{}</a><p>'
_99_30001 = ' <small><a href="http://192.168.1.4/30000"></a></small><a href="http://192.168.1.4/30000/{}192.168.1.4&PassWord">{}192.168.1.4&PassWord</a><p>'  # NOQA


class EthernetRelayMock(object):
    def __init__(self, num_relays=16, init_state=None):
        self.relays = [False] * num_relays
        if init_state:
            assert len(init_state) == num_relays
            self.relays = [s == '1' for s in init_state]

    def get_state_str(self):
        state = ''
        for r in self.relays:
            if r:
                state += '1'
            else:
                state += '0'
        return state

    def toggle(self, relay_index):
        self.relays[relay_index] = not self.relays[relay_index]

    def turn_on(self, relay_index):
        self.relays[relay_index] = True

    def turn_off(self, relay_index):
        self.relays[relay_index] = False

    def all_on(self):
        self.relays = [True] * len(self.relays)

    def all_off(self):
        self.relays = [False] * len(self.relays)


@httmock.all_requests
def mock_relay(url, request):
    """Mocks requests.

    if port is 30000, act normal.
    if port is 30001, act normal but with a messy state.

    if port is 20000, return status_code=200 but bad content

    if port is 40000, return status_code=404
    """
    status_code = 200
    content = None
    match = re.match(r'http://192\.168\.1\.4/(?P<port>\d+)/(?P<num>\d*)', request.url)
    port = int(match.groupdict()['port'])
    num = int(match.groupdict()['num'])
    if port == 30000:
        if num == 99:
            content = _99_30000.format(_MOCK.get_state_str(), _MOCK.get_state_str()).encode('ascii')
        elif num == 44:
            _MOCK.all_off()
        elif num == 45:
            _MOCK.all_on()
        elif num < 40:
            if num % 2 == 1:
                _MOCK.turn_on(int(num / 2))
            else:
                _MOCK.turn_off(int(num / 2))
    elif port == 30001:
        if num == 99:
            content = _99_30001.format(_MOCK.get_state_str(), _MOCK.get_state_str()).encode('ascii')
    elif port == 40000:
        status_code = 404
    elif port == 20000:
        content = 'garbage'.encode('ascii')
    return {'status_code': status_code,
            'content': content}


class TestEthernetRelay(unittest.TestCase):
    """Tests for `sainsmart` package."""

    def setUp(self):
        """Set up test fixtures, if any."""
        global _MOCK
        assert _MOCK is None
        _MOCK = EthernetRelayMock()

    def tearDown(self):
        """Tear down test fixtures, if any."""
        global _MOCK
        _MOCK = None

    def test_000(self):
        """Test init with 30000."""
        with httmock.HTTMock(mock_relay):
            devices.EthernetRelay()

    def test_001(self):
        """Test all on."""
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay()
            relay.all_off()
            relay.all_on()
            self.assertTrue(all(relay.state()))

    def test_002(self):
        """Test all off."""
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay()
            relay.all_on()
            relay.all_off()
            self.assertFalse(any(relay.state()))

    def test_003(self):
        """Test turn on and turn off."""
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay()
            for i in range(len(relay.state())):
                relay.turn_on(i)
            self.assertTrue(all(relay.state()))
            for i in range(len(relay.state())):
                relay.turn_off(i)
            self.assertFalse(any(relay.state()))

    def test_004(self):
        """Test toggle."""
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay()
            for i in range(len(relay.state())):
                relay.toggle(i)
            self.assertTrue(all(relay.state()))
            for i in range(len(relay.state())):
                relay.toggle(i)
            self.assertFalse(any(relay.state()))

    def test_005(self):
        """Non-zero init."""
        global _MOCK
        _MOCK = EthernetRelayMock(num_relays=16, init_state='1010101010101010')
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay()
            self.assertEqual(sum(relay.state()), 8)

    def test_006(self):
        """Test verify."""
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay()
            relay._relays[0] = True  # this should be disallowed anyways
            with self.assertRaises(ValueError):
                relay.verify()

    def test_007(self):
        """Test check_index."""
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay()
            with self.assertRaises(IndexError):
                relay.check_index(-1)
            with self.assertRaises(IndexError):
                relay.check_index(16)

    def test_008(self):
        """Test bad status code."""
        with httmock.HTTMock(mock_relay):
            with self.assertRaises(RuntimeError):
                devices.EthernetRelay(url_base='http://192.168.1.4/40000')

    def test_009(self):
        """Test unable to parse content."""
        with httmock.HTTMock(mock_relay):
            with self.assertRaises(RuntimeError):
                devices.EthernetRelay(url_base='http://192.168.1.4/20000')

    def test_010(self):
        """Test init with 30001."""
        with httmock.HTTMock(mock_relay):
            relay = devices.EthernetRelay(url_base='http://192.168.1.4/30001')
            self.assertTrue(len(relay.state()), 16)


tests/test_sainsmart.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `sainsmart` package."""


import unittest
import httmock
import re

from sainsmart import sainsmart


_MOCK = None
_99 = ' <small><a href="http://192.168.1.4/30000"></a></small><a href="http://192.168.1.4/30000/{}">{}</a><p>'


class EthernetRelayMock(object):
    def __init__(self, num_relays=16, init_state=None):
        self.relays = [False] * num_relays
        if init_state:
            assert len(init_state) == num_relays
            self.relays = [s == '1' for s in init_state]

    def get_state_str(self):
        state = ''
        for r in self.relays:
            if r:
                state += '1'
            else:
                state += '0'
        return state

    def toggle(self, relay_index):
        self.relays[relay_index] = not self.relays[relay_index]

    def turn_on(self, relay_index):
        self.relays[relay_index] = True

    def turn_off(self, relay_index):
        self.relays[relay_index] = False

    def all_on(self):
        self.relays = [True] * len(self.relays)

    def all_off(self):
        self.relays = [False] * len(self.relays)


@httmock.all_requests
def mock_relay(url, request):
    """Mocks requests.

    if port is 30000, act normal.

    if port is 20000, return status_code=200 but bad content

    if port is 40000, return status_code=404
    """
    status_code = 200
    content = None
    match = re.match(r'http://192\.168\.1\.4/(?P<port>\d+)/(?P<num>\d*)', request.url)
    port = int(match.groupdict()['port'])
    num = int(match.groupdict()['num'])
    if port == 30000:
        if num == 99:
            content = _99.format(_MOCK.get_state_str(), _MOCK.get_state_str()).encode('ascii')
        elif num == 44:
            _MOCK.all_off()
        elif num == 45:
            _MOCK.all_on()
        elif num < 40:
            if num % 2 == 1:
                _MOCK.turn_on(int(num / 2))
            else:
                _MOCK.turn_off(int(num / 2))
    elif port == 40000:
        status_code = 404
    elif port == 20000:
        content = 'garbage'.encode('ascii')
    return {'status_code': status_code,
            'content': content}


class TestSainsmart(unittest.TestCase):
    """Tests for `sainsmart` package."""

    def setUp(self):
        """Set up test fixtures, if any."""
        global _MOCK
        assert _MOCK is None
        _MOCK = EthernetRelayMock()

    def tearDown(self):
        """Tear down test fixtures, if any."""
        global _MOCK
        _MOCK = None

    def test_000(self):
        """Test init."""
        with httmock.HTTMock(mock_relay):
            sainsmart.EthernetRelay()

    def test_001(self):
        """Test all on."""
        with httmock.HTTMock(mock_relay):
            relay = sainsmart.EthernetRelay()
            relay.all_off()
            relay.all_on()
            self.assertTrue(all(relay.state()))

    def test_002(self):
        """Test all off."""
        with httmock.HTTMock(mock_relay):
            relay = sainsmart.EthernetRelay()
            relay.all_on()
            relay.all_off()
            self.assertFalse(any(relay.state()))

    def test_003(self):
        """Test turn on and turn off."""
        with httmock.HTTMock(mock_relay):
            relay = sainsmart.EthernetRelay()
            for i in range(len(relay.state())):
                relay.turn_on(i)
            self.assertTrue(all(relay.state()))
            for i in range(len(relay.state())):
                relay.turn_off(i)
            self.assertFalse(any(relay.state()))

    def test_004(self):
        """Test toggle."""
        with httmock.HTTMock(mock_relay):
            relay = sainsmart.EthernetRelay()
            for i in range(len(relay.state())):
                relay.toggle(i)
            self.assertTrue(all(relay.state()))
            for i in range(len(relay.state())):
                relay.toggle(i)
            self.assertFalse(any(relay.state()))

    def test_005(self):
        """Non-zero init."""
        global _MOCK
        _MOCK = EthernetRelayMock(num_relays=16, init_state='1010101010101010')
        with httmock.HTTMock(mock_relay):
            relay = sainsmart.EthernetRelay()
            self.assertEqual(sum(relay.state()), 8)

    def test_006(self):
        """Test verify."""
        with httmock.HTTMock(mock_relay):
            relay = sainsmart.EthernetRelay()
            relay._relays[0] = True  # this should be disallowed anyways
            with self.assertRaises(ValueError):
                relay.verify()

    def test_007(self):
        """Test check_index."""
        with httmock.HTTMock(mock_relay):
            relay = sainsmart.EthernetRelay()
            with self.assertRaises(IndexError):
                relay.check_index(-1)
            with self.assertRaises(IndexError):
                relay.check_index(16)

    def test_008(self):
        """Test bad status code."""
        with httmock.HTTMock(mock_relay):
            with self.assertRaises(RuntimeError):
                sainsmart.EthernetRelay(url_base='http://192.168.1.4/40000')

    def test_009(self):
        """Test unable to parse content."""
        with httmock.HTTMock(mock_relay):
            with self.assertRaises(RuntimeError):
                sainsmart.EthernetRelay(url_base='http://192.168.1.4/20000')


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file."""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except ImportError:
    from urllib.request import urlopen


GITHUB_REPO = 'vicyap/sainsmart'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key.

    Work around keys with incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read().decode())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning."""
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    """Load yaml config file at the given path."""
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    """Save yaml config file at the given path."""
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Put `encrypted_password` into the deploy section of .travis.yml."""
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    """Add a PyPI password to .travis.yml so that Travis can deploy to PyPI.

    Fetch the Travis public key for the repo, and encrypt the PyPI password
    with it before adding, so that only Travis can decrypt and use the PyPI
    password.
    """
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password.encode()))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


sainsmart/sainsmart.py
# -*- coding: utf-8 -*-

"""SainSmart Module."""

import re
import requests

from typing import List  # noqa: F401


class EthernetRelay(object):
    """SainSmart Ethernet Relay.

    <https://www.sainsmart.com/sainsmart-ethernet-control-module-lan-wan-web-server-control-with-rj45-port.html>
    """

    def __init__(self, url_base='http://192.168.1.4/30000'):
        # type: (str) -> None
        """Initalize the class.

        Kwargs:
            url_base (str): The base url of the ethernet relay.

        Attributes:
            url_base (str): The base url of the ethernet relay.
            relays (:obj:`list` of :obj:`bool`): A list of the on/off state of each relay.
        """
        self.url_base = url_base
        self._relays = self.state()

    def check_index(self, relay_index):
        # type: (int) -> None
        """Check that relay_index is valid.

        This method raises an IndexError if `relay_index` is negative or if
        `relay_index` is greater than or equal to the number of relay.

        Args:
            relay_index (int): the relay index.

        Raises:
            IndexError: If `relay_index` is negative or greater than the
                number of relays.
        """
        if relay_index < 0:
            raise IndexError('relay_index={} cannot be negative'.format(relay_index))
        elif relay_index >= len(self._relays):
            raise IndexError('relay_index={} cannot be greater than {}'.format(
                relay_index, len(self._relays)))

    def state(self):
        # type: () -> List[bool]
        """Get the state of the relays.

        Raises:
            RuntimeError: If there was a problem with requesting the url or
                parsing the response.
        """
        r = requests.get('{}/99'.format(self.url_base))
        if r.status_code == 200:
            match = re.search(r'<a.*>(?P<state_bits>[01]+)</a>', r.content.decode('ascii'))
            if match:
                state_bits_str = match.groupdict()['state_bits']
                return [s == '1' for s in state_bits_str]
            else:
                raise RuntimeError("Could not parse the response from the ethernet relay.")
        else:
            raise RuntimeError('The ethernet relay did not respond with status code 200.')

    def verify(self):
        # type: () -> None
        """Verify the state of the relays matches this class instance's state.

        Raises:
            RuntimeError: :func:`EthernetRelay.state()`
            ValueError: If this class instance's state does not match the state
                of the relays.
        """
        state = self.state()
        index = 0
        for i, j in zip(self._relays, state):
            if i != j:
                raise ValueError('Relay at index={} did not match state={}'.format(index, j))
            index += 1

    def toggle(self, relay_index):
        # type: (int) -> None
        """Toggle the state of a relay.

        Args:
            relay_index (int): the relay index to toggle.

        Raises:
            IndexError: :func:`EthernetRelay.check_index`
        """
        self.check_index(relay_index)
        if self._relays[relay_index]:
            self.turn_off(relay_index)
        else:
            self.turn_on(relay_index)

    def turn_on(self, relay_index):
        # type: (int) -> None
        """Turn a relay on.

        Args:
            relay_index (int): the relay index to turn on.

        Raises:
            IndexError: :func:`EthernetRelay.check_index`
        """
        self.check_index(relay_index)
        requests.get('{}/{:02d}'.format(self.url_base, 2 * relay_index + 1))
        self._relays[relay_index] = True
        self.verify()

    def turn_off(self, relay_index):
        # type: (int) -> None
        """Turn a relay off.

        Args:
            relay_index (int): the relay index to turn off.

        Raises:
            IndexError: :func:`EthernetRelay.check_index`
        """
        self.check_index(relay_index)
        requests.get('{}/{:02d}'.format(self.url_base, 2 * relay_index))
        self._relays[relay_index] = False
        self.verify()

    def all_on(self):
        # type: () -> None
        """Turn all relays on."""
        requests.get('{}/45'.format(self.url_base))
        self._relays = [True for r in self._relays]
        self.verify()

    def all_off(self):
        # type: () -> None
        """Turn all relays off."""
        requests.get('{}/44'.format(self.url_base))
        self._relays = [False for r in self._relays]
        self.verify()


sainsmart/devices.py
# -*- coding: utf-8 -*-

"""Devices Module."""

import re
import requests

from typing import List  # noqa: F401


class EthernetRelay(object):
    """SainSmart Ethernet Relay.

    <https://www.sainsmart.com/sainsmart-ethernet-control-module-lan-wan-web-server-control-with-rj45-port.html>
    """

    def __init__(self, url_base='http://192.168.1.4/30000'):
        # type: (str) -> None
        """Initalize the class.

        Kwargs:
            url_base (str): The base url of the ethernet relay.

        Attributes:
            url_base (str): The base url of the ethernet relay.
            relays (:obj:`list` of :obj:`bool`): A list of the on/off state of each relay.
        """
        self.url_base = url_base
        self._relays = self.state()

    def check_index(self, relay_index):
        # type: (int) -> None
        """Check that relay_index is valid.

        This method raises an IndexError if `relay_index` is negative or if
        `relay_index` is greater than or equal to the number of relay.

        Args:
            relay_index (int): the relay index.

        Raises:
            IndexError: If `relay_index` is negative or greater than the
                number of relays.
        """
        if relay_index < 0:
            raise IndexError('relay_index={} cannot be negative'.format(relay_index))
        elif relay_index >= len(self._relays):
            raise IndexError('relay_index={} cannot be greater than {}'.format(
                relay_index, len(self._relays)))

    def state(self):
        # type: () -> List[bool]
        """Get the state of the relays.

        Raises:
            RuntimeError: If there was a problem with requesting the url or
                parsing the response.
        """
        r = requests.get('{}/99'.format(self.url_base))
        if r.status_code == 200:
            content = r.content.decode('ascii')
            re.sub('192.168\.\d+\.\d+', '', content)
            regex = re.compile(
                r'''
                <a.*>
                (?P<state_bits>[01]+)
                .*</a>
                ''',
                re.VERBOSE)
            match = regex.search(content)
            if match:
                state_bits_str = match.groupdict()['state_bits']
                return [s == '1' for s in state_bits_str]
            else:
                raise RuntimeError("Could not parse the response from the ethernet relay.")
        else:
            raise RuntimeError('The ethernet relay did not respond with status code 200.')

    def verify(self):
        # type: () -> None
        """Verify the state of the relays matches this class instance's state.

        Raises:
            RuntimeError: :func:`EthernetRelay.state()`
            ValueError: If this class instance's state does not match the state
                of the relays.
        """
        state = self.state()
        index = 0
        for i, j in zip(self._relays, state):
            if i != j:
                raise ValueError('Relay at index={} did not match state={}'.format(index, j))
            index += 1

    def toggle(self, relay_index):
        # type: (int) -> None
        """Toggle the state of a relay.

        Args:
            relay_index (int): the relay index to toggle.

        Raises:
            IndexError: :func:`EthernetRelay.check_index`
        """
        self.check_index(relay_index)
        if self._relays[relay_index]:
            self.turn_off(relay_index)
        else:
            self.turn_on(relay_index)

    def turn_on(self, relay_index):
        # type: (int) -> None
        """Turn a relay on.

        Args:
            relay_index (int): the relay index to turn on.

        Raises:
            IndexError: :func:`EthernetRelay.check_index`
        """
        self.check_index(relay_index)
        requests.get('{}/{:02d}'.format(self.url_base, 2 * relay_index + 1))
        self._relays[relay_index] = True
        self.verify()

    def turn_off(self, relay_index):
        # type: (int) -> None
        """Turn a relay off.

        Args:
            relay_index (int): the relay index to turn off.

        Raises:
            IndexError: :func:`EthernetRelay.check_index`
        """
        self.check_index(relay_index)
        requests.get('{}/{:02d}'.format(self.url_base, 2 * relay_index))
        self._relays[relay_index] = False
        self.verify()

    def all_on(self):
        # type: () -> None
        """Turn all relays on."""
        requests.get('{}/45'.format(self.url_base))
        self._relays = [True for r in self._relays]
        self.verify()

    def all_off(self):
        # type: () -> None
        """Turn all relays off."""
        requests.get('{}/44'.format(self.url_base))
        self._relays = [False for r in self._relays]
        self.verify()
